sympy/matrices/expressions/diagonal.py

```diff
@@ -1,14 +1,19 @@
 from __future__ import print_function, division
 
 from sympy.matrices.expressions import MatrixExpr
-from sympy.core import S
-
+from sympy.core import S, Eq
+from sympy.functions.special.tensor_functions import KroneckerDelta
 class DiagonalMatrix(MatrixExpr):
     arg = property(lambda self: self.args[0])
     shape = property(lambda self: (self.arg.shape[0], self.arg.shape[0]))
 
     def _entry(self, i, j):
-        return S.Zero if i != j else self.arg[i, 0]
+        eq = Eq(i, j)
+        if eq is S.false:
+            return S.Zero
+        elif eq is S.true:
+            return self.arg[i, i]*KroneckerDelta(i, j)
+        return self.arg[i, j]*KroneckerDelta(i, j)
```

Comment:
It is here that you could return KD(i,j)*self.args[foo] where foo is `i,i`, `j,j` or `i,j`. To keep it as unobtrusive as possible, what you have is fine. (Though if anyone else has an opinion they are free to give it.)